### DISCUSSION TITLE
LoRa peer to peer and LoRaWAN

### DISCUSSION THREAD

**Discussion by skyeee21** [jgromes/RadioLib#972] [Q&A]
I'd like to ask is it possible to integrate both p2p and LoRaWAN firmware in a single board? The goal is after communicating in p2p the node will transmit the received data to the gateway and to TTN. Anyone who can help please? We're using LA66 Dragino LoRa Shield RFM95 
![IMG_20240215_211228](https://github.com/jgromes/RadioLib/assets/158477861/e5c0a032-d218-4a58-82bb-52a251c1d971)
with Arduino Uno. Please share your thoughts. Thank you

**Comment by jgromes** [comment]
I think you will run into some issues. First and foremost, the LoRaWAN stack is quite large in terms of compiled binary size. That's not unexpected, it's a complex protocol. However, it will most likely not fit into Arduino Uno program memory. Also, RAM usage probably exceeds Uno capacity.

However, even if you did switch to a different board (e.g. something based on ESP32), I don't think it would work, as the LoRaWAN stack may change the LoRa modulation parameters in runtime (due to adaptive data rate requests from the LoRaWAN server), breaking your your P2P link. The alternative would be to always set a known configuration before the P2P communication, however, the LoRaWAN stack won't know about the change, thus likely breaking the LoRaWAN link. Overall I'm not sure this is worth the trouble - why not just send to LoRaWAN directly from the first node, instead of jumping through a "repeater"?

**Comment by HeadBoffin** [comment]
The LA66 as supplied is an AT driven module using the ASR6601 which is an ARM Cortex 4 + SX1262 SIP. The Heltec CubeCell range includes support for that for Arduino but I'm not aware of anyone porting it to the LA66. If you can get that going, then you can use RadioLib. Otherwise it's not the board for you.

For dual use, whilst it's not been tested yet, in theory the @StevenCellist's persistence should allow you to do LoRaWAN, save the state, init the radio to do LoRa, do that and then restore the state of LoRaWAN. The only caveat  to that is there is no way you can hear incoming LoRa whilst the radio is setup for LoRaWAN, so you need to stay in LoRa only mode and then just switch over to LoRaWAN to send/receive. But you'll need something other than LA66 unless you want to port the Heltec BSP for Arduino.
